dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+  * `Refs<T>.Instance` is how it is obtained
+    * Also will do validation which will be shared in the future by the workflow worker
+    * `Refs` is in `Temporal` not `Temporal.Workflow`
+  * This is backed by a dynamic proxy
+  * Really old versions of .NET (i.e. would be rarely used by users, < C# 8), don't allow a `static readonly` property on the
+    interface and therefore they have to make their `Ref` elsewhere
+
+## Activity Definition
+
+Common static definition:
+
+```csharp
+namespace MyNamespace;
+
+using Temporal.Activity;
+
+public static class GreetingActivities
+{
+    [Activity(Name = "create_greeting_activity")]
+    public static async Task<string> CreateGreeting(GreetingInfo info)
+    {
+        return $"{info.Salutation}, {info.Name}!";
+    }
+}
+```
+
+Common instance definition:
+
+```csharp
+namespace MyNamespace;
+
+using Temporal.Activity;
+
+public class GreetingActivities
+{
+    public static readonly GreetingActivities Ref = Refs<GreetingActivities>.Instance;
+
+    [Activity(Name = "create_greeting_activity")]
+    public async Task<string> CreateGreetingAsync(GreetingInfo info)
+    {
+        return $"{info.Salutation}, {info.Name}!";
+    }
+}
+```
+
+Notes:
+
+* Activities can be on static or non-static methods
+  * Even though our proxying library requires virtual methods for interception, we're never intercepting, only
+    _referencing_.
+    * This means if someone _does_ invoke an instance method on a `Ref`, it'd call it normally. We can't help this.
+* All activities must have a non-inheritable `Activity` attribute
+  * Optional `Name` positional property can be given. Default is unqualified method name with "Async" trimmed off the
+    end if it returns a `Task`
+* Activity classes can be/implement interfaces
+  * There is no value to this from a Temporal perspective currently since the caller side of activities is not in this
+    phase
+* Activities do not have to be async, they will use the global thread pool otherwise (see worker later)
+* Non-static activity classes are encouraged to set a `static` `readonly` `Ref` property for use by clients
+  * Similar approach as workflows
+  * Unlike workflow validation that checks every public method attribute, this only ensures there is at least one public
+    activity
+  * Not required for static activities because they can just be referenced on the class
+* Any overridden instance activities from base classes or same-named static activities from base classes must have the
+  exact same signature and attribute. This ensures people don't make mistakes accidentally overriding and removing the
+  activity attribute. This also avoids a diamond problem or duplicate problem for statics.
+
+### Activity Contextual Operations
+
+The activity context looks like this:
+
+```csharp
+namespace Temporal.Activity;
+
+public class ActivityContext
+{
+    public static ActivityContext Current { get; }
+
+    public static bool HasCurrent { get; }
```

Comment:
Hrmm, yeah both have benefits. In this case, `HasX` lets you know you can access property `X`.

I now wonder if, instead of a `Current` (so you have to `ActivityContext.Current.Heartbeat()`) I should just make these static (so you `ActivityContext.Heartbeat()`).

I also considered changing the name of this class to just `Activity` but it clashes with `Temporal.Activity`

Reply:
Concur and I am on the same fence. One thing that doesn't come into play is the verbosity - .NET developers definitely don't mind that.

One reason many of these things are instances instead of static is for DI purposes. I considered making a `IActivityContext` interface, but I couldn't have a static `Current` property on it in _all_ versions of .NET we are targeting (because you can't have any non-interface-methods on interfaces in older .NET versions). So that's why I have it like this.